connect_four_game.cpp

```diff
@@ -0,0 +1,164 @@
+#include <bits/stdc++.h>
+using namespace std;
+
+// Below is the implementation of Connect Four game.
+// Follow up Question to be solved next- Algorithm to find the best colum to drop the coin.
+class ConnectFour
+{
+
+private:
+    int totalRows;
+    int totalCols;
+    // board configuration
+    vector<vector<int>> board;
+    //shows the top available spot for each column
+    vector<int> topSpot;
+
+public:
+    ConnectFour(int totalRows, int totalCols)
+    {
+        this->totalRows = totalRows;
+        this->totalCols = totalCols;
+        board.resize(totalRows, vector<int>(totalCols, 0));
+        //bottom spot available for each column
+        topSpot.resize(totalCols, totalRows - 1);
+    }
+
+    int move(int col, int player)
+    {
+        // This function returns
+        // -1 -> for invalid move
+        // 1 -> player doesn't win after this move
+        // int player -> if player wins after this move
+        // Time Complexity - O(1)
+
+        // If that column doesn't exist or that column is filled,
+        // it's an invalid move.
+        if((col < 0) || (col >= totalCols) || (topSpot[col] < 0))
+        {
+            cout << "Not a valid move\n";
+            return -1;
+        }
+        else
+        {
+            board[topSpot[col]][col] = player;
+            topSpot[col]--;
+        }
+
+        if(isWinner(col, player) != 0)
+            return player;
+
+        return 0;
+    }
+
+    bool isWinner(int col, int player)
+    {
+        // Time Complexity - O(1)
+
+        //row num where player placed the ball
+        int row = topSpot[col] + 1;
+
+        //Checking vertically
+        if(isInside(row, col, row+3, col) && (board[row][col] == player) && 
+            (board[row+1][col] == player) && 
+            (board[row+2][col] == player) && 
+            (board[row+3][col] == player))
+        return player;
+
+        //Checking horizontally
+        for(int j = col - 3; j <= col; j++)
+        {
+            if(isInside(row, j, row, j + 3) && 
+                (board[row][j] == player) && 
+                (board[row][j+1] == player) && 
+                (board[row][j+2] == player) && 
+                (board[row][j+3] == player))
+            return player;
+        }
+
+        //Checking diagonally
+        for(int i = row-3, j = col-3; (i <= row) & (j <= col); i++, j++)
+        {
+            if(isInside(i, j, i + 3, j + 3) && 
+                (board[i][j] == player) && 
+                (board[i+1][j+1] == player) && 
+                (board[i+2][j+2] == player) && 
+                (board[i+3][j+3] == player))
+            return player;
+        }
+
+        //Checking anti-diagonally
+        for(int i = row-3, j = col+3; (i <= row ) & (j >= col); i++, j--)

```

Comment:
nit: Consider setting min and max on the loop range. For example i can start from minRow = max(0, row - 3) and go upto maxRow = min(totalRows, row + 1). This way you can eliminate the need to have IsInside method.